Adjust _WKWebExtensionWindow and _WKWebExtensionTab APIs based on feedback.

xeenon opened the pull request:
#### 8987771749cd19e4bfdff5daf2d57a1f276789f9
<pre>
Adjust _WKWebExtensionWindow and _WKWebExtensionTab APIs based on feedback.
<a href="https://webkit.org/b/260156">https://webkit.org/b/260156</a>

Reviewed by Brian Weinstein.

* Added delegate methods for window and tab creation with new helper classes to hold sugested properties
for the new windows and tabs.
* Have all action methods for windows and tabs take a completionHandler so clients that need do do them
async can report when they are complete.
* Renamed isEphemerial to isUsingPrivateBrowsing to be more clear and match common wording.
* Added missing close method, and setter methods for frame and windowState on the window protocol.
* Removed method for isEphemeral on the tab protocol, since tabs are required to be in a private window.
* Added missing activate, duplicate, and setZoomFactor methods on the tab protocol.

* Source/WebKit/Modules/OSX_Private.modulemap: Added new headers.
* Source/WebKit/Modules/iOS_Private.modulemap: Added new headers.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(toImpl): Fix the cases for All and None.
(-[_WKWebExtensionContext didActivateTab:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController didActivateTab:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTab.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTabCreationOptions.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTabCreationOptions.mm: Added.
(-[_WKWebExtensionTabCreationOptions _init]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTabCreationOptionsInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindow.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptions.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptions.mm: Added.
(-[_WKWebExtensionWindowCreationOptions _init]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptionsInternal.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::didActivateTab): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::WebExtensionTab):
(WebKit::WebExtensionTab::window const):
(WebKit::WebExtensionTab::index const): Added.
(WebKit::WebExtensionTab::isPrivate const): Return the result from the containing window.
(WebKit::WebExtensionTab::toggleReaderMode):
(WebKit::WebExtensionTab::mute):
(WebKit::WebExtensionTab::unmute):
(WebKit::WebExtensionTab::detectWebpageLocale):
(WebKit::WebExtensionTab::loadURL):
(WebKit::WebExtensionTab::reload):
(WebKit::WebExtensionTab::reloadFromOrigin):
(WebKit::WebExtensionTab::goBack):
(WebKit::WebExtensionTab::goForward):
(WebKit::WebExtensionTab::activate): Added.
(WebKit::WebExtensionTab::select):
(WebKit::WebExtensionTab::duplicate):
(WebKit::WebExtensionTab::close):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionWindowCocoa.mm:
(WebKit::WebExtensionWindow::WebExtensionWindow):
(WebKit::toAPI): Added.
(WebKit::WebExtensionWindow::setState): Added.
(WebKit::WebExtensionWindow::isPrivate const): Renamed from isEphemeral.
(WebKit::WebExtensionWindow::setFrame): Added.
(WebKit::WebExtensionWindow::close): Added.
(WebKit::WebExtensionWindow::isEphemeral const): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionWindow.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added new files.

Canonical link: <a href="https://commits.webkit.org/267240@main">https://commits.webkit.org/267240@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2c2489a5356ee689660860bfab74c88ab4cc9203

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16074 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/16392 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/16788 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/17838 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15091 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/19405 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/16499 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/17516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/16267 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/16732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/13718 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/18601 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/13975 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/21391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/14964 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/14701 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/17942 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/15300 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/12973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/14532 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/18902 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1968 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15119 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
```diff
@@ -180,12 +198,28 @@
-void WebExtensionTab::toggleReaderMode()
+bool WebExtensionTab::isPrivate() const
+{
+    if (!isValid())
```
Do we want a `m_respondsToIsPrivate` and to check that here?

xeenon replied:
Not needed, since we ask the window and it does the check.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
```diff
@@ -180,12 +198,28 @@
+            return;
+        }
+
+        completionHandler(std::nullopt);
```
Nit: `completionHandler(error ? error.localizedDescription : std::nullopt);`

xeenon replied:
Was going to log the error above, so I'll add that now.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
```diff
@@ -204,20 +238,34 @@
+            return;
+        }
+
+        completionHandler(std::nullopt);
```
Ditto in all of the other places? Unless there is going to be more logic in the error cases.

xeenon replied:
Meant to add logging. Will do that now.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
```diff
@@ -65,15 +65,21 @@ class WebExtensionTab : public RefCounted<WebExtensionTab> {
         URL        = 1 << 8,
         ZoomFactor = 1 << 9,
-        All        = std::numeric_limits<uint16_t>::max(),
+        All        = Audible | Loading | Muted | Pinned | ReaderMode | Size | Title | URL | ZoomFactor,
```
Was this part of a different PR too?

xeenon replied:
Yeah I moved it here, it was in the other one by mistake.

xeenon committed:
Adjust _WKWebExtensionWindow and _WKWebExtensionTab APIs based on feedback.
https://webkit.org/b/260156

Reviewed by Brian Weinstein.

* Added delegate methods for window and tab creation with new helper classes to hold sugested properties
for the new windows and tabs.
* Have all action methods for windows and tabs take a completionHandler so clients that need do do them
async can report when they are complete.
* Renamed isEphemerial to isUsingPrivateBrowsing to be more clear and match common wording.
* Added missing close method, and setter methods for frame and windowState on the window protocol.
* Removed method for isEphemeral on the tab protocol, since tabs are required to be in a private window.
* Added missing activate, duplicate, and setZoomFactor methods on the tab protocol.

* Source/WebKit/Modules/OSX_Private.modulemap: Added new headers.
* Source/WebKit/Modules/iOS_Private.modulemap: Added new headers.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(toImpl): Fix the cases for All and None.
(-[_WKWebExtensionContext didActivateTab:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionController.mm:
(-[_WKWebExtensionController didActivateTab:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionControllerDelegate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTab.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTabCreationOptions.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTabCreationOptions.mm: Added.
(-[_WKWebExtensionTabCreationOptions _init]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionTabCreationOptionsInternal.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindow.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptions.h: Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptions.mm: Added.
(-[_WKWebExtensionWindowCreationOptions _init]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionWindowCreationOptionsInternal.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::didActivateTab): Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionTabCocoa.mm:
(WebKit::WebExtensionTab::WebExtensionTab):
(WebKit::WebExtensionTab::window const):
(WebKit::WebExtensionTab::index const): Added.
(WebKit::WebExtensionTab::isPrivate const): Return the result from the containing window.
(WebKit::WebExtensionTab::toggleReaderMode):
(WebKit::WebExtensionTab::mute):
(WebKit::WebExtensionTab::unmute):
(WebKit::WebExtensionTab::detectWebpageLocale):
(WebKit::WebExtensionTab::loadURL):
(WebKit::WebExtensionTab::reload):
(WebKit::WebExtensionTab::reloadFromOrigin):
(WebKit::WebExtensionTab::goBack):
(WebKit::WebExtensionTab::goForward):
(WebKit::WebExtensionTab::activate): Added.
(WebKit::WebExtensionTab::select):
(WebKit::WebExtensionTab::duplicate):
(WebKit::WebExtensionTab::close):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionWindowCocoa.mm:
(WebKit::WebExtensionWindow::WebExtensionWindow):
(WebKit::toAPI): Added.
(WebKit::WebExtensionWindow::setState): Added.
(WebKit::WebExtensionWindow::isPrivate const): Renamed from isEphemeral.
(WebKit::WebExtensionWindow::setFrame): Added.
(WebKit::WebExtensionWindow::close): Added.
(WebKit::WebExtensionWindow::isEphemeral const): Deleted.
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionWindow.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj: Added new files.

Canonical link: https://commits.webkit.org/267240@main
